website/notifications/events/model.py

```diff
@@ -0,0 +1,352 @@
+# -*- coding: utf-8 -*-
+"""Basic Event handling for events that need subscriptions"""
+
+from furl import furl
+from datetime import datetime
+from six import add_metaclass
+
+from website.notifications import emails
+from website.notifications.constants import NOTIFICATION_TYPES
+from website.models import Node
+from website.notifications.events import utils as event_utils
+from website.notifications import utils
+
+
+class _EventMeta(type):
+    registry = {}
+
+    def __init__(cls, name, bases, attrs):
+        if name != 'Event':
+            event_id = name.lower()
+            cls.registry[event_id] = cls
+
+        super(_EventMeta, cls).__init__(name, bases, attrs)
+
+
+@add_metaclass(_EventMeta)
+class Event(object):
+    """
+    Base notification class for building notification events and messages.
+    - abstract methods set methods that must be defined by subclasses.
+
+    To use this interface you must use the class as a Super (inherited).
+     - Implement property methods in subclasses
+     - All subclasses must be in this file for the meta class to list them
+     - Name the subclasses you will be calling as such:
+      - event (the type of event from _SUBSCRIPTIONS_AVAILABLE or specific cases)
+      - class
+      example: event = file_added, class = FileAdded
+    """
+
+    def __init__(self, user, node, action):
+        self.user = user
+        self.gravatar_url = user.gravatar_url
+        self.node = node
+        self.action = action
+        self.timestamp = datetime.utcnow()
+
+    @classmethod
+    def parse_event(cls, user, node, event, **kwargs):
+        """If there is a class that matches the event then it returns that instance"""
+        kind = ''.join(event.split('_'))
```

Comment:
Because the actions coming in from waterbutler are not camel case, and have underlines. So unless we move away from the usual class names or modify the events from waterbutler this seems to be a compromise. However, if there is a simpler, less magic way please explain.